docs/rollover.adoc

```diff
@@ -0,0 +1,65 @@
+
+== `rollover.hpp`
+
+https://github.com/intel/cpp-std-extensions/blob/main/include/stdx/rollover.hpp[`rollover.hpp`]
+provides a class template `rollover_t` that is intended to act like an unsigned
+integral type, but with semantics that include the ability to roll-over on overflow.
+
+A `rollover_t` can be instantiated with any unsigned integral type:
+
+[source,cpp]
+----
+// explicit type
+auto x = stdx::rollover_t<std::uint8_t>{};
+
+// deduced type is equivalent to the unsigned version of the argument
+auto y = stdx::rollover_t{1}; // rollover_t<unsigned int>
+----
+
+It supports all the usual arithmetic operations (`+` `-` `*` `/` `%`) and
+behaves much like an unsigned integral type, with defined overflow and underflow
+semantics. The difference is in the comparison operators. A `rollover_t`
+considers itself to be in the middle of a rolling window where half the
+bit-range is always lower and half is higher.
+
+For instance, imagine a 3-bit unsigned integral type. There are eight values of
+this type: `0` `1` `2` `3` `4` `5` `6` `7`. For any value, there are always four
+values less than it, and three values greater than it. (And of course it is equal to
+itself, the eighth value). e.g. for the value `5`:
+
+- `1`, `2`, `3`, `4` are all less than `5`
+- `6`, `7`, `0` are all greater than `5`
+
+Effectively any value partitions the cyclic space into (almost) equal parts.
+
+CAUTION: `operator<` on `rollover_t` is not antisymmetric!
+
+This has some interesting consequences, for example both of the following are true:
+
+- `1` < `5`
+- but also, `5` < `1`! (The four values less than `1` are `5`, `6`, `7`, `0`)
```

Comment:
thought: this doesn't seem right to me....is this due to off-by one? it seems like this should not happen.

Reply:
I agree it is uncomfortable, but having thought about it I think it is just a special case of transitivity. I don't think it can be avoided, it's just a property of the rollover nature.

Reply:
I think this might only be avoided by making the partitioned space unequal. So introducing an off-by-N error or making some values incomparable perhaps?